### PR TITLE
Add parser from netlink structs to protobuf

### PR DESCRIPTION
**Pull request by gfr10598**
This change is [Reviewable](https://reviewable.io/reviews/m-lab/tcpinfo_lib/6)

### CODE REVIEW

**Comment by gfr10598** [comment]
There is a lot of interface that is still not fleshed out.  This is just providing the basic parsing functionality.

TODO: 
  There are a few places where errors are not yet handled gracefully.

File: src/tcpinfo.proto [hunk 1]
```diff
@@ -254,17 +255,23 @@ message TCPDiagnosticsProto {
 
   // The following three are mutually exclusive, as they provide
   // data from different congestion control strategies.
-  // Data obtained from struct tcpvegas_info.
-  optional TCPVegasInfoProto vegas     = 4;
-  // Data obtained from struct tcp_dctcp_info.
-  optional DCTCPInfoProto dctcp        = 5;
-  // Data obtained from struct tcp_bbr_info.
-  optional BBRInfoProto bbr_info       = 6;
+  oneof cc_info {
```
Comment (pboothe) [inline src/tcpinfo.proto:258]: Nice use of oneof.

File: src/tcpinfo_lib.cc [hunk 2]
```diff
@@ -14,8 +14,207 @@
...
+#include <linux/unix_diag.h>
+
 namespace mlab {
 namespace netlink {
+namespace {
+unsigned int extract(const std::string& ip, int n) {
+  return (((unsigned char)(ip[n])) << 8) + (unsigned char)ip[n+1];
+}
+
+void ParseInetDiagMsg(struct inet_diag_msg* r, InetDiagMsgProto* proto) {
```
Comment (pboothe) [inline src/tcpinfo_lib.cc:37]: It looks like `proto` is never null in practice, but I don't know enough about the macros you are using to determine if `r` might ever be `NULL` (or `std::nullptr`, I suppose).

File: src/tcpinfo_lib.cc [hunk 3]
```diff
@@ -14,8 +14,207 @@
...
+  proto->set_expires(r->idiag_expires);
+  if (r->idiag_rqueue)
+    proto->set_rqueue(r->idiag_rqueue);
+  if (r->idiag_wqueue)
+    proto->set_wqueue(r->idiag_wqueue);
+  proto->set_uid(r->idiag_uid);
+  proto->set_inode(r->idiag_inode);
+}
+
+void ParseBBRInfo(const struct rtattr* rta, BBRInfoProto* proto) {
```
Comment (pboothe) [inline src/tcpinfo_lib.cc:81]: Same `NULL` worries.

File: src/tcpinfo_lib.cc [hunk 4]
```diff
@@ -14,8 +14,207 @@
...
+  proto->set_rmem_alloc(info->idiag_rmem);
+  proto->set_wmem_alloc(info->idiag_wmem);
+  proto->set_fwd_alloc(info->idiag_fmem);
+  proto->set_tmem(info->idiag_tmem);
+}
+
+void ParseSKMemInfo(const struct rtattr* tb,
+                  SocketMemInfoProto* proto) {
+  const __u32* info = (__u32*)RTA_DATA(tb);
+#define SET_NONZERO(tag, field) \
```
Comment (pboothe) [inline src/tcpinfo_lib.cc]: Please choose a different name? I recognize that you engaged in best practices and undefined the macro above, but it's still weird to see a macro have two different meanings in so short a time.

File: src/tcpinfo_lib.cc [hunk 5]
```diff
@@ -14,8 +14,207 @@
...
+            (unsigned char)ip[2], (unsigned char)ip[3], ep.port());
+  } else if (ip.size() == 16) {
+    sprintf(
+        result, "%x:%x:%x:%x:%x:%x:%x:%x:%u",
+        extract(ip, 0), extract(ip, 2), extract(ip, 4), extract(ip, 6),
+        extract(ip, 8), extract(ip, 10), extract(ip, 12), extract(ip, 14),
+            ep.port());
+  } else {
+    // TODO - should log an error.
+    return "Unrecognized address size.";
```
Comment (pboothe) [inline src/tcpinfo_lib.cc:234]: Maybe return the empty string?

File: src/tcpinfo_lib.cc [hunk 6]
```diff
@@ -24,8 +223,57 @@ TCPDiagnosticsProto TCPInfoParser::ParseNLMsg(
...
+
+  // This may be overwritten below.
+  proto->set_diag_protocol(protocol);
+
+  auto* rta = (struct rtattr *)(r+1);
+  int len = nlh->nlmsg_len - NLMSG_LENGTH(sizeof(*r));
+  while (RTA_OK(rta, len)) {
+    switch (rta->rta_type) {
+      case INET_DIAG_PROTOCOL:
+        proto->set_diag_protocol(Protocol(*(__u8 *)RTA_DATA(rta)));
```
Comment (pboothe) [inline src/tcpinfo_lib.cc:263]: I would be more comfortable if we logged a message if `Protocol(*(__u8 *)RTA_DATA(rta)) != protocol`

File: src/tcpinfo_lib.cc [hunk 7]
```diff
@@ -24,8 +223,57 @@ TCPDiagnosticsProto TCPInfoParser::ParseNLMsg(
...
+          ParseMemInfo(rta, proto->mutable_socket_mem());
+        }
+      case INET_DIAG_BBRINFO:
+        ParseBBRInfo(rta, proto->mutable_bbr_info());
+        break;
+      case INET_DIAG_VEGASINFO:
+        fprintf(stderr, "Need to do vegas\n");
+        break;
+      case INET_DIAG_SKV6ONLY:
+        // TODO(gfr) Do we need this?
```
Comment (pboothe) [inline src/tcpinfo_lib.cc:289]: I have no idea what it is...

File: src/tcpinfo_util.h [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  unsigned int uid;
+  int       refcnt;
+  unsigned int      iface;
+  unsigned long long  sk;
+  char *name;
+  char *peer_name;
+  __u32       mark;
+};
+
+#ifdef __cplusplus
```
Comment (pboothe) [inline src/tcpinfo_util.h]: Do we need to add `#ifndef EXTERN_C` ? This is a pretty common idiom...

**Review by pboothe — COMMENTED** [review]
I have passed over the CMake stuff because I don't know it. It looks sensible, but I will take it on faith that it works. Little worries about checking against null, particularly for code that is `extern`. Other than that, things look good.

File: src/tcpinfo_lib.cc (hunk 5 above)
Comment (gfr10598) [reply src/tcpinfo_lib.cc:234]: This should never happen, so I think having an informative string is ok.

File: src/tcpinfo_lib.cc (hunk 6 above)
Comment (gfr10598) [reply src/tcpinfo_lib.cc:263]: Adding a TODO to add logging later.

File: src/tcpinfo_lib.cc (hunk 7 above)
Comment (gfr10598) [reply src/tcpinfo_lib.cc:289]: Me neither, hence the TODO.  8-)

File: src/tcpinfo_util.h (hunk 8 above)
Comment (gfr10598) [reply src/tcpinfo_util.h]: Good idea.  Thanks!

Comment (gfr10598) [reply src/tcpinfo_util.h]: Also, this file shouldn't be in this PR.  Removed.

**Review by pboothe — APPROVED** [review]
LGTM

**Pull request merged**